Objects/typeobject.c

```diff
@@ -6599,6 +6599,12 @@ object_set_class(PyObject *self, PyObject *value, void *closure)
     }
 
     PyTypeObject *oldto = Py_TYPE(self);
+#ifdef Py_GIL_DISABLED
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    // The real Py_TYPE(self) (`oldto`) may have changed from
+    // underneath us in another thread, so we re-fetch it here.
+    _PyEval_StopTheWorld(interp);
```

Comment:
This should be moved up one line `PyTypeObject *oldto = Py_TYPE(self);` because the type might change across the stop the world call.